### DISCUSSION TITLE
Domes, making them work

### DISCUSSION THREAD

**Discussion by mefistotelis** [swfans/syndwarsfx#171] [General]
> That is something I thought of suggesting as well - an enhancement so that domes have no collision when fully opened. It's fairly clear they just don't work properly in the final game and were abandoned as a device. If they could actually be entered when opened then all kinds of interesting things could be built around it, like hiding items in there, or vehicles. The scripted sequence with the soldier in the dome in the mission The Dead Zone would actually look right then.
> I actually wonder why they didn't get this working, they're clearly an old game element that goes back to the pre-alpha, but ended up useless - unless they did get it working but it caused other issues?

Re-computing the routing triangulation of an area based on 3D geometry intersecting the ground is kind of taxing in regard to CPU cycles. There was a noticeable slowdown when a building ends collapsing and the triangulation below switches flags to be passable, and that part was especially optimized.

So I guess that was the issue - the pathfinding in pre-alpha was simpler and could handle that in real-time, but when they made 'proper' pathfinding that become problematic.

For opening the dome, simplified procedure could be used- as for collapsing.
But then closing it, would require expensive re-computation, or a different kind of smart optimization.
Shouldn't be that much of a problem today though.

_Originally discussed by @mefistotelis and @Moburma in https://github.com/swfans/swars/discussions/170#discussioncomment-9534902_

---

Creating a new discussion from it, as PAN is not exactly the same subject.

**Comment by mefistotelis** [comment]
Now looking at the details of the domes, it would seem the devs did tried to make it work again - they just didn't had the time to debug complex problem.

As usually in such cases - if something doesn't want to work easily, there's probably more than one bug preventing that.

For now, I found one bug - the opening/closing speed was increased 4x compared to pre-alpha, and that meant skipping some animation frames (3 out of 4, to be exact). But - change in passability was supposed to happen at a specific frame, and that specific frame was now skipped.

After fixing that, the domes still do not allow to be entered - so there is clearly something more wrong, somewhere.

**Reply by Moburma** [reply]
I guess it must have been a fairly small window that domes actually worked - if they ever did. The pre-alpha they have no collision at all, and in the beta demo they work as in the final game. Makes sense that little was done with them in the campaign, as they must have known they were broken and not going to be fixed from quite early on (early 1996?). 

I wonder also if the Bahrain multiplayer level was cut (it was in the older multi .miss file) because the dome was broken. It's a rather unfinished level either way.

The Dead Zone level is interesting too as the briefing has a nonsense video showing something else instead of what clearly was going to be the dome sequence. Maybe that was the last chance to get domes working (because they thought the idea of the soldier imprisoned in the dome was going to look cool), then they failed and the left the placeholder/deliberately wrong video in as they never had a chance to record the real working one. It's one of the last levels worked on in August, so maybe that was the dome's final failed hurrah as a concept.

**Comment by mefistotelis** [comment]
Now I see the remaining issue: BuildStartVect and BuildNumbVect are not set. They're zero.

Either they're on a different position and need updating on map load, or they're gone and need to be re-created.

(these are position and count of `ColVectList` items assigned to the geometry of an Object)

For the Grav Gun dome in map `-m 99,16`, the range should include values 1396-1420 (taken from debug collisions `-d c` mode)
@Moburma would you be able to verify if the thing linked to this dome has such value somewhere?
BuildStartVect should be somewhere in range <1100, 1396>, and BuildNumbVect in range <24, 100>.

**Reply by Moburma** [reply]
Do you mean the Thing for the dome in the map file?

**Reply by mefistotelis** [reply]
yes.

**Reply by Moburma** [reply]
Sorry, had to go write a script to dump it out. Output:
https://file.io/QfzKB5j2egjP

So, the dome is Thing 287 according to the open command

For that Thing I see 
BuildStartVect =  1378
BuildNumbVect = 43

So it **is** populated in the map file?

**Reply by mefistotelis** [reply]
Definitely did not expected that. Will check where the value disappears then.

When this is fixed, wouldn't that make `-m 1,85` harder? Can the player open the dome on this level, or do you have to sync your rescue with one of the weapon tests?

**Comment by Moburma** [comment]
That's a very good question. Do we actually know how it will respond to the
person being inside a closed dome, I assume they will remain trapped
inside? It could be they get pushed out by the collision or something!

We could easily add in an invisible npc to perform an open command on the
dome if the agent who does the "testing" is dead.

There's also the possibility of the player getting stuck inside a dome in
e.g the Audacity levels, but you'd have to actively try to have that
occur...

**Reply by mefistotelis** [reply]
> Do we actually know how it will respond to the person being inside a closed dome

Try this:

```
diff --git a/src/lvfiles.c b/src/lvfiles.c
index e67e764..bbf4db3 100644
--- a/src/lvfiles.c
+++ b/src/lvfiles.c
@@ -682,6 +682,7 @@ void load_level_pc(short level, short missi, ubyte reload)
         if (level_deep_fix) {
             level_perform_deep_fix();
         }
+        reset_things_col_vect_range();
         fix_level_indexes(missi, fmtver, reload, level_deep_fix);
     } else
     {
```

**Comment by mefistotelis** [comment]
I now understand the issue.

The game does not copy all properties from Objects in map file, instead it is very selective in what it gets.

The function making the selection is `new_thing_building_clone()`.

We probably want to add copying more properties there.

EDIT:

Ok I guess we should modify the levels to make dome opening easier, and then we can enable entering them.

The proper place to enable is now here:
```
diff --git a/src/thing.c b/src/thing.c
index 0b1e134..80ed5bd 100644
--- a/src/thing.c
+++ b/src/thing.c
@@ -853,7 +853,7 @@ ThingIdx new_thing_building_clone(struct Thing *p_clthing, struct M33 *p_clmat,
         p_thing->U.UObject.PrevThing = p_clthing->U.UObject.PrevThing;
         p_thing->U.UObject.OffX = p_clthing->U.UObject.OffX;
         p_thing->U.UObject.OffZ = p_clthing->U.UObject.OffZ;
-#if 0 // Disabled until the domes entering is considered safe to enable
+#if 1 // Disabled until the domes entering is considered safe to enable
         p_thing->U.UObject.BuildStartVect = p_clthing->U.UObject.BuildStartVect;
         p_thing->U.UObject.BuildNumbVect = p_clthing->U.UObject.BuildNumbVect;
 #endif
```

**Reply by Moburma** [reply]
Nice, I've been out all day but had a quick play around now, seems to work well. The only thing is the collisions on the side of the domes feels a bit off, agents will try and walk through these if you click anywhere near them. Also this interrupts the "tester" from shooting the soldier on the Dead Zone level, which looks goofy (but not as stupid as it used to look!). 

Dead Zone definitely needs a failsafe to open the dome now, I will do this later today, got some things to do first (won't take long).

Is anyone aware of any other levels that actually do anything with domes that is worth testing? All I can think of is the pointless open and closing on the Audacity levels.

**Reply by Moburma** [reply]
Here's the version with the dome opener, seems to work fine:
https://filebin.net/ezrakugi2yt57g46/C003L002.DAT

**Reply by mefistotelis** [reply]
Open on death works.
Will persuading the agent open the dome as well?

**Reply by Moburma** [reply]
Not sure. I had hoped to test it, but embarrassingly even with cheats I keep failing to achieve this! Will try again later. 

EDIT:
One new thing I noticed, in the Audacity etc levels the policemen at the domes now actually go inside them and regard the err... things inside them. What they are meant to be and why they do this? No one knows.

**Reply by mefistotelis** [reply]
>  even with cheats I keep failing to achieve this!

Maybe our 'beef up' cheat should also give you persuasion power?

**Reply by Moburma** [reply]
Yes, I was thinking that. This level is a massive ballache just to get enough PP (and keep it) to persuade the target, let alone agents as well. Even running around with maximum health I'd get all my persuaded people wiped out by satelllite rain or bombs or whatever.

**Reply by mefistotelis** [reply]
Updated the cheats. Doesn't look like persuasion counts - the dome remains closed.

**Reply by Moburma** [reply]
I'm over-thinking it - probably easiest just to have a generic group 0 arrive check on box around the dome to cause it to always open, rather then bother checking for dead npcs etc.

**Reply by mefistotelis** [reply]
yep, that would work.

**Reply by Moburma** [reply]
This just opens the dome when anyone from your group is near:
https://filebin.net/5z1jp90x2kcph5on

EDIT: oops, forgot to hide my man. New version!

**Reply by mefistotelis** [reply]
Works fine. Not sure if it should close again when group 0 is away.. but regardless - works fine. We can push that to the levels repo.

(I never realized you can antagonize male vs female Ungiuded on this level. That clears a substantial part of the punks.)

**Reply by Moburma** [reply]
Yeah, I didn't close it because this is a pretty annoying level in general and thought it best to give the player an easy ride at the end. The open command doesn't loop, so in theory you could open the dome, then have the agent come and close it again, but God only knows how you would manage that as you'd pretty much have to entirely leave the base to avoid him due to his patrol route, so it's not really a risk.

@geist22 pointed out the ability to make the unguided fight on this level, I always love the ability to do stuff like that in games.

**Reply by geist22** [reply]
I believe there are 3 different Unguided groups on this level. I've went through a lot of pain during my most recent max rewards challenge run trying to carry 16 High Explosives out of this mission. Once you kill (or attack?) a member of a group, ALL members of this group on the entire map will start moving towards your agents to engage them. The groups aren't allied, so they can hit each other on accident and will start fighting. One group has everyone with a Minigun, the second group roams mostly in the middle, has Pulse Laser/Electron Maces and everyone carries an explosive. The third group seems to be rather small with a few male Unguided with Launchers and LR Rifles, usually on the other side of the map near the IML/bank.

**Comment by mefistotelis** [comment]
> the collisions on the side of the domes feels a bit off, agents will try and walk through these if you click anywhere near them

Yeah, and note it doesn't happen if you do `reset_things_col_vect_range()` fix.
So it looks like the values within things are set, but are slightly off.

We will still need to reset the values on every level load.

**Reply by mefistotelis** [reply]
No, the values were not off, just had different meaning than I assumed. The range should only contain the switchable part, not the whole range associated with the thing.

After adjusting my code to such definition, the `reset_things_col_vect_range()` is no longer required.

**Comment by mefistotelis** [comment]
A bit strange that the dome opening/closing doesn't make any sound.

Do we have any samples which would match?

EDIT:
Feature commited - the domes can now be entered when opened.

**Reply by Moburma** [reply]
Not sure really, the standard door open/close, or the tank turret noise?

**Reply by mefistotelis** [reply]
The door sounds are a bit too short.
And the tank turret sounds.. too 'alarming'?

I don't know, I'm not good with sounds. But we should invent something.
I imagine something like a idle motorcycle sound - then we could repeat it for the duration of opening/closing.

Some inspiration:
https://www.youtube.com/watch?v=9bgFpss0YFE

**Reply by Moburma** [reply]
Yeah, I will have a think...
On a kind of tangent, I made this a while ago:
https://pastebin.com/rThbP7Gn
It's a list of what all known sound slots actually do. For all we know there may already be one for domes opening that is just not populated. Theme Hospital works the same way, there are sounds set to play in the final game that don't as nothing is in that slot in the sound.dat.

EDIT: Also the first thing that came to mind was actually the Doom door open sound!

EDIT2: Duh, that video made me realise the SW .wav file annotations actually tell you which sound effects collection they are taken from. Like the person getting hit sounds specifically say they are from Sound Ideas 6032.
Another example, the Displacertron has this comment: "HB03 51 - 33 seconds to go"
Which leads us to this:
https://archive.org/details/99WindHowlingWhistlingWindWeather/51+Sci+Fi+Space+Scatter+Ray.mp3

**Reply by mefistotelis** [reply]
We have the whole dome state control function, it's `process_dome1`.
No sample playback is called there.

**Reply by Moburma** [reply]
6 months later, I finally had a proper crack at this. See what you think:
https://fastupload.io/8c0ed3ee87b3569a

It's made up of a sample of a garage door opener mixed with someone sliding a baking tray out of an oven.

I might:

- stretch the baking tray sound to last longer over the whole sample - I like the idea of the metal plates in the dome scraping together as they move
- Get rid of the squeaky motor sound at the end. I liked it at first, but think it may drag hearing it over time. Let me know what you think
- If I do get rid of the end sound, I might just put some kind of metal clanging/boom of it coming into place

I think it's about the right length.

EDIT:

Here's with the baking tray sound stretched out, think that is better:

https://fastupload.io/30d83a2d51f1c5e6

**Reply by mefistotelis** [reply]
The link opens a new ad window every time I click something, but no download.

**Reply by Moburma** [reply]
Ok, here's another host:
Original
https://easyupload.io/ayc3ud
Improved
https://easyupload.io/111nkc

**Reply by Moburma** [reply]
Actually, how about this:
https://easyupload.io/5kpdo4

Got rid of the stupid end bit (not sure why I thought that was good). Most of the other sounds in the game are fairly clean and simplistic, so think this is "enough". What do people think, good, bad?

**Reply by mefistotelis** [reply]
Sounds good, We'll see how it matches in the game.

Not sure if it should be one sample or more like two - one at start of the transformation, and one just before it's done. Two samples would better match the game regardless of time taken by opening. Ie. one sample could be played until opening is 90% done, then the sample would be stopped and another one played.

So this is opening, are you able to prepare closing as well?

**Reply by Moburma** [reply]
> So this is opening, are you able to prepare closing as well?

How about this:

https://easyupload.io/piw50h

Think the final sound is still a bit too loud/in your face, but do we like the general idea?

Two samples is a good idea, but not sure how I would do that yet.

**Reply by mefistotelis** [reply]
That sounds good to me.

> Two samples is a good idea, but not sure how I would do that yet.

We should probably make one first and start thinking about two if it the length of sample to animation will really be a problem.

**Reply by mefistotelis** [reply]
I added playing some samples to `process_dome1()`.
Now you should be able to easily replace the sample numbers with yours, to test the solution.
Sound banks will get auto-generated if you create a PR for `swars-sfx` which adds the sounds to filelist.

EDIT:
I promise I did not really selected the sound I used, just copied one, turns out from the tank. But.. that sounds better as final sample of the closing animation than as the tank sound. Maybe we should find a better match for the tank rotation sound, and leave this one for the dome?

**Reply by Moburma** [reply]
Hmm, it's not too bad, although a bit loud. I'll have a play about with it.

**Reply by Moburma** [reply]
How about this:

https://www.youtube.com/watch?v=4AZigyinMjs

EDIT:

Stretched the sound out a bit more, fits nicely now

https://www.youtube.com/watch?v=7tsZUNm7FEM

**Reply by mefistotelis** [reply]
That sounds good to me.

**Reply by Moburma** [reply]
I think I'll commit this now, if no one disagrees. I used slot 45 in the sound.dat for the new sound I made. I'll upload it to the sfx repository and add it to the files.

I tried making a closing sound as well, but it didn't sound that great, I think just re-using the same sound as in the above videos works fine.

**Comment by Moburma** [comment]
Also on the subject of domes, -m 99,16 is clearly not finished but could be spruced up. It's pretty apparent the leather jacket men throughout are supposed to be invisible, so that would be an easy win. So too there was likely going to be a second item in the other dome, as explained above there is an unused trigger for it. 
But what should it be?

I guess this is all academic as multiplayer doesn't work, but interesting to think of.

**Reply by mefistotelis** [reply]
> It's pretty apparent the leather jacket men throughout are supposed to be invisible,

These invisible men are surprising to me. i mean, on many levels they do some job and then just stay idle. But they do consume a Thing slot, and their state is processed every turn. So they're a waste of memory and CPU, even if marginal.

I would just ping them into existence and turn into normal civilians (give a command to wander).

> I guess this is all academic as multiplayer doesn't work

Sooner or later someone interested in MP will find this place.
And the task isn't that complex - KeeperFX now has TCP-IP support, so it's mostly a matter of copying the solution.

**Comment by geist22** [comment]
Good stuff. I was a bit sceptical at first, since this makes some changes to Church mission 18 and I'm naturally resistant to gameplay changes, though the prisoner still survives up to the Nuclear Grenade.

My suggestion would be to lower the proximity at which your Zealots will cause the dome to open to about the same of the agent or slightly lower, opening it from too far feels off to me. A good distance I think would be the border between the bright texture from the hallway and the dark texture around the dome. A higher distance feels like the dome just opens automatically rather than your Zealots opening it. Too "artificial", you know?

The dome staying open forever if your Zealots open it also doesn't feel right to me, it closing again when you leave the trigger area would feel more natural. If it's not a one time event you could always come back and open it again in case a persuaded follower got stuck there, right?

**Reply by Moburma** [reply]
> the prisoner still survives up to the Nuclear Grenade.

Something funny I realised while working on this - he's actually invincible. You can't kill him, and neither can the agent. What actually happens is he has a check setup to see if the _dome_ is still intact or not. When the agent blows the dome up, he sees it is gone, and kills himself! So in fact this is a suicide prevention mission. This is also why you have to persuade him, his commands are all around blowing himself up, so you persuade him to interrupt this and make him follow.

I might change the dome opening area as you say to make the cause and effect a little more obvious. I agree closing it again would be neater, but I don't want people getting stuck/him getting stuck in it - this is one of the hardest levels in the entire game and I've certainly rage quit it myself before even when the dome wasn't in there, so I'd rather keep it foolproof and not have this possibly go wrong.
